server/routerlicious/packages/protocol-base/src/protocol.ts

```diff
@@ -24,23 +25,41 @@ export interface IScribeProtocolState {
     values: [string, ICommittedProposal][];
 }
 
-export function isSystemMessage(message: ISequencedDocumentMessage) {
+/**
+ * Ops that client (container runtime & Container class) generates.
+ * Everything else is a service op
+ */
+export function isClientMessage(message: ISequencedDocumentMessage | IDocumentMessage) {
     switch (message.type) {
-        case MessageType.ClientJoin:
-        case MessageType.ClientLeave:
         case MessageType.Propose:
         case MessageType.Reject:
         case MessageType.NoOp:
-        case MessageType.NoClient:
         case MessageType.Summarize:
-        case MessageType.SummaryAck:
-        case MessageType.SummaryNack:
+        case MessageType.Operation:
             return true;
         default:
             return false;
     }
 }
 
+/**
+ * Tells if op is not representing modification of user content in container
+ * It includes such things as
+ * - Service generates ops (join, leave, noClient, summary ack/nack)
+ * - Some ops submitted by client, with service being recipient:
+ *    - Summary op submitted by container runtime
+ *    - Protocol (Quorum) ops: propose, reject, Noop
+ *    - Collab control: Noop
+ */
+export function isSystemMessage(message: ISequencedDocumentMessage) {
```

Comment:
>isSystemMessage [](http://example.com/codeflow?start=16&length=15)

Sorry if I'm missing something - isSystemMessage is now only restricted to Operation ? Will it break the existing locations where we consume isSystemMessage ?  Like the new one we added on packages\drivers\odsp-driver\src\fetchSnapshot.ts ? 
 nonSysOps:  snapshot.ops?.filter((op) => !isSystemMessage(op)).length ?? 0,

Reply:
At the moment this implementation is almost equivalent to what we had. The difference (if you look at enum) are remoteHelp (not used), tripComplete (not used) & Control (used only by service internally, never sequenced).

I've opened https://github.com/microsoft/FluidFramework/issues/9709 to cleanup first two.

The change affects future. If service adds new types of ops, then we would not account here. I think it's not wise to limit client to understand only existing ops (especially that some code like loader will be in prod for years to come). So, I want to make sure any new type of ops are treated as system.

This puts more pressure on client not to introduce new types of ops. And that's the reason I'm adding check for isClientMessage when we sent op, and Container.submitContainerMessage() checks that only "op" can come from container runtime.

Reply:
I'd only add that it's more obvious now (with my changes) that name of this API is wrong and needs to change. But I'm leaving it for future :) Actually, maybe we should remove it altogether, given that it's trivial to inline this check.

Reply:
Please double check me here (on logic) that it's equivalent other than those 3 cases :)

Reply:
Yes, they are the same - got it :)  I wasn't sure these 3 types were not going to matter .